client-js/app/elements/labrad-nodes.ts

```diff
@@ -331,152 +392,227 @@ export class LabradNodes extends polymer.Base {
+                                            : 'Version Conflict!';
+
+      const isGlobal = server.environmentVars.length === 0;
+      const newServer = {
+        name: server.name,
+        version: version,
+        autostart: false,
+        errorString: '',
+        errorException: '',
+        nodes: []
+      };
+
+      if (isGlobal) {
+        const idx = this.getServerIndex(server, this.globalServers);
+        if (idx === -1) {
+          this.push('globalServers', newServer);
+          if (filterFunction(newServer)) {
+            this.push('globalServersFiltered', newServer);
+          }
+        }
+      } else {
+        const idx = this.getServerIndex(server, this.localServers);
+        if (idx === -1) {
+          this.push('localServers', newServer);
+          if (filterFunction(newServer)) {
+            this.push('localServersFiltered', newServer);
+          }
         }
       }
     }
-    var names = Array.from(nameSet.values());
-    names.sort();
-    return names;
+
+    this.updateNodeServerBinding('globalServersFiltered');
+    this.updateNodeServerBinding('localServersFiltered');
+
+    this.updateFilters();
+
+    this.kick++;
   }
 
-  private _statusMap(info: Array<NodeStatus>): Map<string, Map<string, ServerStatus>> {
-    var statusMap = new Map<string, Map<string, ServerStatus>>();
-    for (let nodeStatus of info) {
-      var serverMap = new Map<string, ServerStatus>();
-      for (let serverStatus of nodeStatus.servers) {
-        serverMap.set(serverStatus.name, serverStatus);
+  private removeItemFromList(item: ServerDisconnectMessage): void {
+    const idx = this.getNodeIndex(item, this.info);
+    if (idx !== -1) {
+      this.splice('info', idx, 1);
+    }
+
+    this.updateNodeServerBinding('globalServersFiltered');
+    this.updateNodeServerBinding('localServersFiltered');
+
+    this.updateFilters();
+
+    this.kick++;
+  }
+
+
+  private _serverStatusMap(statuses: ServerStatus[]): Map<String, ServerStatus> {
+    const serverStatusMap = new Map<String, ServerStatus>();
+    for (const s of statuses) {
+      serverStatusMap.set(s.name, s);
+    }
+    return serverStatusMap;
+  }
+
+
+  private _nodeAutostartSet(node: NodeStatus): Set<String> {
+    const nodeAutostartSet = new Set();
+    for (const server of node.autostartList) {
+      nodeAutostartSet.add(server);
+    }
+    return nodeAutostartSet;
+  }
+
+  private removeOfflineNodes(serversName: string,
+                             serverIdx: number): void {
+    const onlineNodes = new Set();
+    for (const node of this.info) {
+      onlineNodes.add(node.name);
+    }
+
+    // Backwards as we are splicing from the list as we iterate.
+    const nodes = this[serversName][serverIdx].nodes;
+    for (let nodeIdx = nodes.length - 1; nodeIdx >= 0; --nodeIdx) {
+      const node = nodes[nodeIdx];
+      if (!onlineNodes.has(node.name)) {
+        this.splice(`${serversName}.#${serverIdx}.nodes`, nodeIdx, 1);
```

Comment:
A comment about what is going on here would be helpful. That `#` is rather cryptic.